File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.java
package com.virtuslab.gitmachete.frontend.actions.dialogs;

import java.awt.Dimension;
import java.awt.GridLayout;

import javax.swing.BoxLayout;
import javax.swing.JComponent;
import javax.swing.JLabel;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogPanel;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.ui.components.JBTextField;
import org.checkerframework.checker.guieffect.qual.UIEffect;
import org.checkerframework.checker.nullness.qual.Nullable;

import com.virtuslab.gitmachete.frontend.actions.common.GitMacheteBundle;

public final class SlideInDialog extends DialogWrapper {

  private static final int COMPONENT_WIDTH = 360;

  private final JBTextField textField = new JBTextField();

  @UIEffect
  public SlideInDialog(Project project, String parentName) {
    super(project);
    setTitle(GitMacheteBundle.message("action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.title", parentName));

    init();
    setOKButtonText(GitMacheteBundle.message("action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.ok-button"));
    setOKButtonMnemonic('I');
  }

  @Override
  @UIEffect
  protected JComponent createCenterPanel() {
    var label = new JLabel(GitMacheteBundle.message("action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.label"));
    var panel = new DialogPanel();
    panel.setLayout(new GridLayout(/* rows */ 2, /* cols */ 1));
    panel.setPreferredSize(new Dimension(COMPONENT_WIDTH, textField.getHeight()));
    panel.add(row(label));
    panel.add(row(textField));
    panel.setPreferredFocusedComponent(textField);
    return panel;
  }

  @Nullable
  @UIEffect
  public String showAndGetEntryName() {
    return showAndGet() ? textField.getText().trim() : null;
  }

  @UIEffect
  private static JComponent row(JComponent component) {
    component.setLayout(new BoxLayout(component, BoxLayout.X_AXIS));
    return component;
  }

}

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt
package com.virtuslab.gitmachete.frontend.actions.dialogs

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.ui.DocumentAdapter
import com.intellij.ui.layout.CellBuilder
import com.intellij.ui.layout.ValidationInfoBuilder
import com.intellij.ui.layout.panel
import com.virtuslab.branchlayout.api.IBranchLayout
import com.virtuslab.branchlayout.api.IBranchLayoutEntry
import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle.getString
import java.awt.event.KeyEvent
import javax.swing.JCheckBox
import kotlin.apply
import kotlin.text.isEmpty
import kotlin.text.trim

data class SlideInOptions(
    val name: String,
    @get:JvmName("shouldReattach")
    val reattach: Boolean = true)

class SlideInDialog
    constructor(
        project: Project, private val branchLayout: IBranchLayout, private val parentName: String
    ) : DialogWrapper(project, /* canBeParent */ true) {

  // this field is only ever meant to be written on UI thread
  private var branchName = ""
  private var reattach = false
  private var reattachCheckbox: JCheckBox? = null
  private val rootNames = branchLayout.rootEntries.map { it.name }

  init {
    title = getString("action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.title")
    setOKButtonText(
        getString("action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.ok-button"))
    setOKButtonMnemonic('I'.toInt())
    init()
  }

  fun showAndGetBranchName() =
      if (showAndGet()) SlideInOptions(branchName.trim(), reattach) else null

  override fun createCenterPanel() =
      panel {
        row(
            getString(
                "action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.label.parent")) {
          label(parentName, bold = true)
        }
        row {
          label(
              getString(
                  "action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.label.branch-name"))
        }
        row {
          textField(::branchName, { branchName = it })
              .focused()
              .withValidationOnApply(validateBranchName())
              .apply { startTrackingValidationIfNeeded() }
        }
        row {
          reattachCheckbox =
              checkBox(
                  getString(
                      "action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.checkbox.reattach"),
                  ::reattach)
                  .component
                  .apply {
                mnemonic = KeyEvent.VK_R
                isEnabled = false
                isSelected = false
              }
        }
      }

  private fun validateBranchName():
      ValidationInfoBuilder.(javax.swing.JTextField) -> ValidationInfo? =
      {
        val errorInfo = git4idea.validators.checkRefName(it.text)
        if (errorInfo != null) error(errorInfo.message)
        else if (it.text == parentName) {
          error(
              getString(
                  "action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.error.slide-in-under-itself"))
        } else {
          val entryByName = branchLayout.findEntryByName(it.text)
          if (entryByName
              .map { isDescendant(presumedDescendantName = parentName, presumedAncestorEntry = it) }
              .getOrElse(false)) {
            error(
                getString(
                    "action.GitMachete.BaseSlideInBranchBelowAction.dialog.slide-in.error.slide-in-under-its-descendant"))
          } else {
            if (rootNames.contains(it.text)) { // the provided branch name refers to the root entry
              reattachCheckbox?.isEnabled = false
              reattachCheckbox?.isSelected = true
            } else {
              val existsAndHasAChild = entryByName.orNull?.children?.nonEmpty() ?: false
              reattachCheckbox?.isEnabled = existsAndHasAChild
              reattachCheckbox?.isSelected =
                  reattachCheckbox?.isSelected ?: false && existsAndHasAChild
            }

            null
          }
        }
      }

  private fun CellBuilder<javax.swing.JTextField>.startTrackingValidationIfNeeded() {
    if (branchName.isEmpty()) {
      component.document
          .addDocumentListener(
              object : DocumentAdapter() {
                override fun textChanged(e: javax.swing.event.DocumentEvent) {
                  startTrackingValidation()
                  component.document.removeDocumentListener(this)
                }
              })
    } else {
      startTrackingValidation()
    }
  }

  private fun isDescendant(
      presumedDescendantName: String, presumedAncestorEntry: IBranchLayoutEntry
  ): Boolean {
    if (presumedAncestorEntry.children.exists { it.name.equals(presumedDescendantName) }) {
      return true
    } else {
      return presumedAncestorEntry.children.exists() { isDescendant(presumedDescendantName, it) }
    }
  }
}
